### ISSUE TITLE
Concurrency bug(s) in LogWriter

### ISSUE THREAD

**Issue by mathewc** [Azure/azure-webjobs-sdk#1313]
Seeing the following stack trace in our error logs, indicating a concurrency issue:

```
System.InvalidOperationException : Collection was modified; 
enumeration operation may not execute.
   at System.Collections.Generic.HashSet`1.Enumerator.MoveNext()
   at async Microsoft.Azure.WebJobs.Logging.LogWriter.FlushIntancesAsync()
   at async Microsoft.Azure.WebJobs.Logging.LogWriter.FlushCoreAsync()
```

There are several HashSets used in this class. Need to review them all and see where we're not locking properly.

**Comment by mathewc** [comment]
Related to https://github.com/Azure/azure-webjobs-sdk-script/issues/1403?

**Comment by brettsam** [comment]
You beat me to it :-)
 @MikeStall -- It looks like your fix for Azure/azure-webjobs-sdk-script#1403 takes care of this as well (you are already locking around the `foreach`). This exception is different, but that's just b/c it's a race to see which side throws first (enumerating or adding). The lock around the `Add` call should take care of this.
